### PR TITLE
Create `QuillToolbarSelectHeaderStyleDropdownButton`

### PR DESCRIPTION
**Pull request by MacDeveloper1**
## Description

The alternative widget `QuillToolbarSelectHeaderStyleDropdownButton` for heading style is created. Also the list of available headings is expanded to H6. So the new widget supports headings H1..H6. Font sizes,line heights and spacing are updated from this [link](https://boosted.orange.com/docs/4.6/content/typography/). Basically standard HTML values are used.

P.S. Also some translations (uk and ru) are updated.

## Images

image
image
image

## Checklist

- [x] I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [ ] I titled the PR using [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0).
- [x] I did not modify the `CHANGELOG.md` nor the plugin version in `pubspec.yaml` files.
- [ ] I have run the commands in `./scripts/before-push.sh` and it all passed successfully

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate that with a `!` in the title as explained in [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0)).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by freshtechtips** [comment]
> ## Description
> The alternative widget `QuillToolbarSelectHeaderStyleDropdownButton` for heading style is created. Also the list of available headings is expanded to H6. So the new widget supports headings H1..H6. Font sizes,line heights and spacing are updated from this [link](https://boosted.orange.com/docs/4.6/content/typography/). Basically standard HTML values are used.
> 
> P.S. Also some translations (uk and ru) are updated.
> 
> ## Images
> image image image
> 
> ## Checklist
> * [x]  I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
> * [ ]  I titled the PR using [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0).
> * [x]  I did not modify the `CHANGELOG.md` nor the plugin version in `pubspec.yaml` files.
> * [ ]  I have run the commands in `./scripts/before-push.sh` and it all passed successfully
> 
> ## Breaking Change
> Does your PR require plugin users to manually update their apps to accommodate your change?
> 
> * [ ]  Yes, this is a breaking change (please indicate that with a `!` in the title as explained in [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0)).
> * [x]  No, this is _not_ a breaking change.

Hi, I appreciate your efforts but we already did that in the dev branch, we will look into some solution don't worry we will try to merge your PR

**Comment by freshtechtips** [comment]
> ## Description
> 
> The alternative widget `QuillToolbarSelectHeaderStyleDropdownButton` for heading style is created. Also the list of available headings is expanded to H6. So the new widget supports headings H1..H6. Font sizes,line heights and spacing are updated from this [link](https://boosted.orange.com/docs/4.6/content/typography/). Basically standard HTML values are used.
> 
> P.S. Also some translations (uk and ru) are updated.
> 
> ## Images
> 
> image
> image
> image
> 
> 
> ## Checklist
> 
> - [x] I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
> - [ ] I titled the PR using [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0).
> - [x] I did not modify the `CHANGELOG.md` nor the plugin version in `pubspec.yaml` files.
> - [ ] I have run the commands in `./scripts/before-push.sh` and it all passed successfully
> 
> ## Breaking Change
> 
> Does your PR require plugin users to manually update their apps to accommodate your change?
> 
> - [ ] Yes, this is a breaking change (please indicate that with a `!` in the title as explained in [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0)).
> - [x] No, this is *not* a breaking change.

Can you fix Ci failure by doing it aginst the latest master branch?

**Comment by freshtechtips** [comment]
Besides the header styles for the new ones
We need to ask @singerdmx first because that's up this library is a port of react js so it needs to be the same otherwise quill delta will not work on quill js

And that's the whole idea behind the repo or the port

**Comment by MacDeveloper1** [comment]
I tested this new class and it works. Qulll does not care on how many H is specified in configuration. Configuration is customizable.
Please look at [this](https://stackoverflow.com/questions/56845622/include-headings-in-quill-rich-editor).

**Comment by freshtechtips** [comment]
> @freshtechtips  and @singerdmx I tested this new class and it works. See images. All headings are shown correctly. Qulll does not care on how many H is specified in configuration. Configuration is customizable.
> Please look at [this](https://stackoverflow.com/questions/56845622/include-headings-in-quill-rich-editor).

I'm not sure but I think you are right, but  can at least do it with the new class that was added before you sent this PR but to the dev branch, and now it's merged with the master

**Comment by MacDeveloper1** [comment]
The example code which I've just took from [this](https://quilljs.com/docs/modules/toolbar/).
```js
var toolbarOptions = [
  ['bold', 'italic', 'underline', 'strike'],        // toggled buttons
  ['blockquote', 'code-block'],

  [{ 'header': 1 }, { 'header': 2 }],               // custom button values
  [{ 'list': 'ordered'}, { 'list': 'bullet' }],
  [{ 'script': 'sub'}, { 'script': 'super' }],      // superscript/subscript
  [{ 'indent': '-1'}, { 'indent': '+1' }],          // outdent/indent
  [{ 'direction': 'rtl' }],                         // text direction

  [{ 'size': ['small', false, 'large', 'huge'] }],  // custom dropdown
  [{ 'header': [1, 2, 3, 4, 5, 6, false] }],  // <------------------------------ headers

  [{ 'color': [] }, { 'background': [] }],          // dropdown with defaults from theme
  [{ 'font': [] }],
  [{ 'align': [] }],

  ['clean']                                         // remove formatting button
];

var quill = new Quill('#editor', {
  modules: {
    toolbar: toolbarOptions
  },
  theme: 'snow'
});
```

The aim of Quill I think is to keep rich text which can be converted to HTML which in its turn supports H1...H6.

**Comment by MacDeveloper1** [comment]
> but can at least do it with the new class that was added before you sent this PR but to the dev branch, and now it's merged with the master

It seems I didn't catch what do you mean.

**Comment by freshtechtips** [comment]
> The example code which I've just took from [this](https://quilljs.com/docs/modules/toolbar/).
> 
> ```js
> var toolbarOptions = [
>   ['bold', 'italic', 'underline', 'strike'],        // toggled buttons
>   ['blockquote', 'code-block'],
> 
>   [{ 'header': 1 }, { 'header': 2 }],               // custom button values
>   [{ 'list': 'ordered'}, { 'list': 'bullet' }],
>   [{ 'script': 'sub'}, { 'script': 'super' }],      // superscript/subscript
>   [{ 'indent': '-1'}, { 'indent': '+1' }],          // outdent/indent
>   [{ 'direction': 'rtl' }],                         // text direction
> 
>   [{ 'size': ['small', false, 'large', 'huge'] }],  // custom dropdown
>   [{ 'header': [1, 2, 3, 4, 5, 6, false] }],  // <------------------------------ headers
> 
>   [{ 'color': [] }, { 'background': [] }],          // dropdown with defaults from theme
>   [{ 'font': [] }],
>   [{ 'align': [] }],
> 
>   ['clean']                                         // remove formatting button
> ];
> 
> var quill = new Quill('#editor', {
>   modules: {
>     toolbar: toolbarOptions
>   },
>   theme: 'snow'
> });
> ```
> 
> The aim of Quill I think is to keep rich text which can be converted to HTML which in its turn supports H1...H6.

Sure but that's not the problem, the problem we already have restored the deleted button from the past in 2022, I restored it with fewer features, and yours contains more features, do you want to fix the conflicts or let me handle it?

**Comment by MacDeveloper1** [comment]
> the problem we already have restored the deleted button from the past in 2022, I restored it with fewer features, and yours contains more features, do you want to fix the conflicts or let me handle it?

I will appreciate you if you can do this. And I think I could not do it by myself because of `Only those with [write access] to this repository can merge pull requests`. `Resolve conflicts` button is disabled for me.

**Comment by MacDeveloper1** [comment]
I think there are not many places of default package code which must be changed. The new widget and config are new files. It is required to update Attributes to add new ones from H4 to H6 and expand `switch` statements to handle new headers.

**Comment by freshtechtips** [comment]
> I think there are not many places of default package code which must be changed. The new widget and config are new files.
> 
> It is required to update Attributes to add new ones from H4 to H6 and expand `switch` statements to handle new headers. Also `DefaultStyles` are updated for H1...H6.

Do you want me to carry the changes to a separate PR or fetch and edit this PR? What do you prefer?

**Comment by MacDeveloper1** [comment]
> Do you want me to carry the changes to a separate PR or fetch and edit this PR? What do you prefer? if you want me to choose I would choose the first because of the too many conflicts
Please do as you feel comfortable.

**Comment by freshtechtips** [comment]
> > Do you want me to carry the changes to a separate PR or fetch and edit this PR? What do you prefer? if you want me to choose I would choose the first because of the too many conflicts
> 
> Please do as you feel comfortable. The important thing is correct work of my app which already uses this new widget my own branch. :)

Just one thing, is the header styles always up to h6 from h1 or the developer can override it?

File: lib/src/widgets/default_styles.dart [hunk 1]
```diff
@@ -220,47 +250,95 @@ class DefaultStyles {
...
           defaultTextStyle.style.copyWith(
-            fontSize: 20,
-            color: defaultTextStyle.style.color!.withOpacity(0.70),
-            height: 1.25,
-            fontWeight: FontWeight.w500,
+            fontSize: 24,
+            color: defaultTextStyle.style.color,
+            letterSpacing: -0.5,
+            height: 1.083,
+            fontWeight: FontWeight.bold,
```
Comment (freshtechtips) [inline lib/src/widgets/default_styles.dart:280]: I think there is a bug in those changes

**Comment by freshtechtips** [comment]
Did you face this bug?
https://github.com/singerdmx/flutter-quill/assets/73608287/330c66ab-2c2e-436d-bc4e-3d413ca50e38

**Comment by freshtechtips** [comment]
This pull request has been moved to another, but it has a reference

**Comment by MacDeveloper1** [comment]
> Just one thing, is the header styles always up to h6 from h1 or the developer can override it?

It is fixed from H1 to H6 and then Normal. I thought it is logical. It is shown in the image I post in start topic.

File: lib/src/l10n/generated/quill_localizations.dart [hunk 2]
```diff
@@ -636,6 +636,48 @@ abstract class FlutterQuillLocalizations {
...
+
+  /// No description provided for @normal.
+  ///
+  /// In en, this message translates to:
+  /// **'Normal'**
+  String get normal => 'Normal';
+
+  /// No description provided for @heading1.
+  ///
+  /// In en, this message translates to:
```
[the remark this answers is not in the document]
Comment (MacDeveloper1) [reply lib/src/l10n/generated/quill_localizations.dart:648]: Sorry. I was to hurry. I think I can update translations after merge the PR with new widget.

File: lib/src/l10n/generated/quill_localizations_ru.dart [hunk 3]
```diff
@@ -229,22 +229,43 @@ class FlutterQuillLocalizationsRu extends FlutterQuillLocalizations {
...
+  String get pasteAPhotoUsingALink => 'Вставьте фотографию, используя ссылку';
 
   @override
-  String get pickAVideoFromYourGallery => 'Pick a video from your gallery';
+  String get pickAVideoFromYourGallery => 'Выберете видео из вашей галереи';
 
   @override
   String get recordAVideoUsingYourCamera =>
-      'Record a video using your phone camera';
+      'Запишите видео, используя вдеокамеру';
```
[the remark this answers is not in the document]
Comment (MacDeveloper1) [reply lib/src/l10n/generated/quill_localizations_ru.dart:247]: Sure.

File: lib/src/widgets/default_styles.dart [hunk 4]
```diff
@@ -220,47 +250,95 @@ class DefaultStyles {
         h1: DefaultTextBlockStyle(
             defaultTextStyle.style.copyWith(
               fontSize: 34,
-              color: defaultTextStyle.style.color!.withOpacity(0.70),
-              height: 1.15,
-              fontWeight: FontWeight.w300,
+              color: defaultTextStyle.style.color,
+              letterSpacing: -1,
+              height: 1,
+              fontWeight: FontWeight.bold,
```
[the remark this answers is not in the document]
Comment (MacDeveloper1) [reply lib/src/widgets/default_styles.dart:256]: As I added new heading I need to have som wide used values for heading so I searched the Internet and find a few links which suggest such values. The link with data which I used you may find in the start post. 
Usually headers are used for highlighting some titles but opacity do the reading hard making usual text more higlighted then header. Usually defautt values for text black or white depending on theme brightness.

File: lib/src/widgets/default_styles.dart [hunk 5]
```diff
@@ -220,47 +250,95 @@ class DefaultStyles {
...
             defaultTextStyle.style.copyWith(
-              fontSize: 24,
-              color: defaultTextStyle.style.color!.withOpacity(0.70),
-              height: 1.15,
-              fontWeight: FontWeight.normal,
+              fontSize: 30,
+              color: defaultTextStyle.style.color,
+              letterSpacing: -0.8,
+              height: 1.067,
+              fontWeight: FontWeight.bold,
```
[the remark this answers is not in the document]
Comment (MacDeveloper1) [reply lib/src/widgets/default_styles.dart:268]: I took the data from [here](https://boosted.orange.com/docs/4.6/content/typography/)

> Get started with Boosted, a Bootstrap fork — the world’s most popular framework for building responsive, mobile-first sites, with jsDelivr and a template starter page.

.

**Comment by freshtechtips** [comment]
> > Just one thing, is the header styles always up to h6 from h1 or the developer can override it?
> 
> It is fixed list from H1 to H6 and then Normal. I thought it is logical. It is shown in the image I post in start topic.

I add the ability to edit it

**Comment by MacDeveloper1** [comment]
> I add the ability to edit it

Why do think it is necessary? To place more large font somewhere in middle? Or do you mean moving only _Normal_ options to top or bottom?

**Comment by freshtechtips** [comment]
> > I add the ability to edit it
> 
> Why do think it is necessary? To place more large font somewhere in middle? Or do you mean moving only _Normal_ options to top or bottom?
> 

No, it's just one variable, which is a list that has default value, and you can change it back to previous value, it's necessary otherwise this can be considered as a breaking change

**Comment by freshtechtips** [comment]
We already carried the changes of this PR, thank you

**Pull request closed without merging**